src/gdb/GDBDebugSessionBase.ts

```diff
@@ -623,6 +645,52 @@ export abstract class GDBDebugSessionBase extends LoggingDebugSession {
         }
     }
 
+    /**
+     * Pause thread if running and suppress stopped event.
+     * In contrast to pauseIfNeeded(), there is no expectation to resume after a debugger operation.
+     *
+     * Notes:
+     *   - NOT to be called during configuration phase.
+     *   - No requireAsync, supposed to be removed from pauseIfNeeded in future.
+     */
+    protected async pauseIfRunning(): Promise<void> {
+        if (!this.isRunning) {
+            // All threads paused already
+            return;
+        }
+        // Get current thread ID
+        const currentThreadId = await this.gdb.queryCurrentThreadId();
+        if (this.gdb.isNonStopMode()) {
+            // this.isRunning means at least one thread is running. Check if current one.
+            const currentThread = this.threads.filter(
+                (thread) => thread.id === currentThreadId
+            );
+            if (currentThread.every((thread) => !thread.running)) {
+                // Current thread already paused
+                return;
+            }
+        }
```

Comment:
The comments here are incorrect, so the early-return logic can be simplified: `isRunning` means “all threads are running” (with a special case for no threads), not “any threads are running” – see `updateIsRunning`.

```suggestion
        if (!this.isRunning) {
            // Some threads paused already
            return;
        }
```

Note that technically there are race conditions here, but they are not new, so don’t need fixing in this PR: `isRunning` reflects the last state that GDB has confirmed to us, but what we should be checking here is the last state that we have asked GDB for. If `isRunning` is true but another pause request is already in flight, we don’t need another one; conversely if a continue request is in flight but we haven’t received confirmation of its effect yet so `isRunning` is still false, we do need a pause request.